event-handler/app.go

```diff
@@ -0,0 +1,415 @@
+	app.semaphore = make(chan struct{}, c.Concurrency)
+	app.sessions = make(chan session)
+
+	return app, nil
+}
+
+// Run initializes and runs a watcher and a callback server
+func (a *App) Run(ctx context.Context) error {
+	a.Process = lib.NewProcess(ctx)
+
+	a.SpawnCriticalJob(a.mainJob)
+	a.SpawnCriticalJob(a.sessionConsumerJob)
+	<-a.Process.Done()
+
+	return a.Err()
+}
+
+// Err returns the error app finished with.
+func (a *App) Err() error {
+	return trace.Wrap(a.mainJob.Err())
+}
+
+// WaitReady waits for http and watcher service to start up.
+func (a *App) WaitReady(ctx context.Context) (bool, error) {
+	mainReady, err := a.mainJob.WaitReady(ctx)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+
+	sessionConsuerReady, err := a.sessionConsumerJob.WaitReady(ctx)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+
+	return mainReady && sessionConsuerReady, nil
+}
+
+// takeSemaphore obtains semaphore
+func (a *App) takeSemaphore(ctx context.Context) error {
+	select {
+	case a.semaphore <- struct{}{}:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// releaseSemaphore releases semaphore
+func (a *App) releaseSemaphore(ctx context.Context) error {
+	select {
+	case <-a.semaphore:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// consumeSession ingests session
+func (a *App) consumeSession(ctx context.Context, s session) error {
+	log := logger.Get(ctx)
+
+	log.WithField("id", s.ID).Info("Started session events ingest")
+
+	url := a.config.FluentdSessionURL + "." + s.ID + ".log"
+
+	chEvt, chErr := a.teleport.StreamSessionEvents(ctx, s.ID, int64(s.Index))
+
+Out:
+	for {
+		select {
+		case err := <-chErr:
+			return trace.Wrap(err)
+
+		case evt := <-chEvt:
+			if evt == nil {
+				log.WithField("id", s.ID).Info("Finished session events ingest")
+
+				// Session export has finished, we do not need it's state anymore
+				err := a.state.RemoveSession(s.ID)
+				if err != nil {
+					return trace.Wrap(err)
+				}
+
+				break Out
+			}
+
+			e, err := NewTeleportEvent(evt, "")
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			_, ok := a.config.SkipSessionTypes[e.Type]
+			if !ok {
+				err := a.sendEvent(ctx, url, &e)
+				if err != nil {
+					return trace.Wrap(err)
+				}
+			}
+
+			// Set session index
+			err = a.state.SetSessionIndex(s.ID, e.Index)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+
+	err := a.state.RemoveSession(s.ID)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	a.releaseSemaphore(ctx)
+
+	return nil
+}
+
+// runSessionConsumer runs session consuming process
+func (a *App) runSessionConsumer(ctx context.Context) error {
+	log := logger.Get(ctx)
+
+	a.sessionConsumerJob.SetReady(true)
+
+	for {
+		select {
+		case s := <-a.sessions:
+			log.WithField("id", s.ID).WithField("index", s.Index).Info("Starting session ingest")
+
+			a.takeSemaphore(ctx)
+			func(s session) {
```

Comment:
Is there a `go` missing here?

Reply:
No, SpawnCritical spawns passed function as a goroutine. go here would be redundant.